File: backend_python/app/services/portfolio_service.py
from collections import defaultdict

from sqlalchemy.orm import Session, joinedload

from app.models.instrument import UserInstrument
from app.models.investment import Investment, InvestmentType
from app.schemas.report import (
    InvestmentTypeBreakdown,
    LotRead,
    PortfolioPlatformBreakdown,
    PortfolioPosition,
    PortfolioReport,
)


def get_portfolio(db: Session, user_id: int) -> PortfolioReport:
    investments = (
        db.query(Investment)
        .options(
            joinedload(Investment.user_instrument).joinedload(UserInstrument.instrument),
            joinedload(Investment.platform_account),
        )
        .filter(
            Investment.user_id == user_id,
            Investment.user_instrument_id.isnot(None),
        )
        .order_by(Investment.purchase_date.asc())
        .all()
    )

    # Group lots by user_instrument_id
    by_instrument: dict[int, list[Investment]] = defaultdict(list)
    for inv in investments:
        by_instrument[inv.user_instrument_id].append(inv)

    positions: list[PortfolioPosition] = []
    by_type: dict[InvestmentType, dict] = defaultdict(lambda: {"invested": 0.0, "current": 0.0, "count": 0})
    by_platform: dict[str, dict] = defaultdict(lambda: {"invested": 0.0, "current": 0.0})

    for ui_id, lots in by_instrument.items():
        first = lots[0]
        instrument = first.user_instrument.instrument
        inv_type = first.type

        total_invested = sum(float(l.amount_invested) for l in lots)
        current_value = sum(
            float(l.current_value) if l.current_value is not None else float(l.amount_invested)
            for l in lots
        )
        unrealized_gain = current_value - total_invested
        unrealized_gain_pct = (unrealized_gain / total_invested * 100) if total_invested else 0.0

        # Units (stock → quantity, MF → units)
        if inv_type == InvestmentType.stock:
            total_units_vals = [float(l.quantity) for l in lots if l.quantity is not None]
            total_units = sum(total_units_vals) if total_units_vals else None
            # Weighted avg buy price: sum(buy_price * qty) / sum(qty)
            weighted = sum(
                float(l.buy_price) * float(l.quantity)
                for l in lots
                if l.buy_price is not None and l.quantity is not None
            )
            qty_sum = sum(float(l.quantity) for l in lots if l.quantity is not None)
            avg_buy_price = weighted / qty_sum if qty_sum else None
        elif inv_type == InvestmentType.mutual_fund:
            total_units_vals = [float(l.units) for l in lots if l.units is not None]
            total_units = sum(total_units_vals) if total_units_vals else None
            # Weighted avg NAV
            weighted = sum(
                float(l.nav_at_purchase) * float(l.units)
                for l in lots
                if l.nav_at_purchase is not None and l.units is not None
            )
            unit_sum = sum(float(l.units) for l in lots if l.units is not None)
            avg_buy_price = weighted / unit_sum if unit_sum else None
        else:
            total_units = None
            avg_buy_price = None

        # Platform accounts holding this position
        platform_names: list[str] = []
        for lot in lots:
            if lot.platform_account:
                name = lot.platform_account.nickname
                if name not in platform_names:
                    platform_names.append(name)
                # Aggregate by_platform
                pa_key = lot.platform_account.nickname
                by_platform[pa_key]["invested"] += float(lot.amount_invested)
                cv = float(lot.current_value) if lot.current_value is not None else float(lot.amount_invested)
                by_platform[pa_key]["current"] += cv

        lot_reads = [
            LotRead(
                id=l.id,
                purchase_date=l.purchase_date,
                amount_invested=float(l.amount_invested),
                current_value=float(l.current_value) if l.current_value is not None else None,
                quantity=float(l.quantity) if l.quantity is not None else None,
                buy_price=float(l.buy_price) if l.buy_price is not None else None,
                folio_number=l.folio_number,
                units=float(l.units) if l.units is not None else None,
                nav_at_purchase=float(l.nav_at_purchase) if l.nav_at_purchase is not None else None,
                platform_account_nickname=l.platform_account.nickname if l.platform_account else None,
                notes=l.notes,
            )
            for l in lots
        ]

        positions.append(
            PortfolioPosition(
                user_instrument_id=ui_id,
                instrument_name=instrument.name,
                instrument_ticker=instrument.ticker_symbol,
                instrument_exchange=instrument.exchange,
                type=inv_type,
                platform_accounts=platform_names,
                total_lots=len(lots),
                total_units=total_units,
                total_invested=total_invested,
                avg_buy_price=avg_buy_price,
                current_value=current_value,
                unrealized_gain=unrealized_gain,
                unrealized_gain_pct=unrealized_gain_pct,
                lots=lot_reads,
            )
        )

        by_type[inv_type]["invested"] += total_invested
        by_type[inv_type]["current"] += current_value
        by_type[inv_type]["count"] += len(lots)

    # Sort: by type (stock first), then by current_value desc within type
    type_order = {InvestmentType.stock: 0, InvestmentType.mutual_fund: 1}
    positions.sort(key=lambda p: (type_order.get(p.type, 99), -p.current_value))

    total_invested_all = sum(p.total_invested for p in positions)
    current_value_all = sum(p.current_value for p in positions)
    unrealized_all = current_value_all - total_invested_all
    unrealized_pct_all = (unrealized_all / total_invested_all * 100) if total_invested_all else 0.0

    type_breakdown = [
        InvestmentTypeBreakdown(
            type=t,
            total_invested=d["invested"],
            current_value=d["current"],
            unrealized_gain=d["current"] - d["invested"],
            count=d["count"],
        )
        for t, d in sorted(by_type.items(), key=lambda x: x[1]["current"], reverse=True)
    ]

    platform_breakdown = sorted(
        [
            PortfolioPlatformBreakdown(
                platform_name=name,
                total_invested=d["invested"],
                current_value=d["current"],
            )
            for name, d in by_platform.items()
        ],
        key=lambda x: x.current_value,
        reverse=True,
    )

    return PortfolioReport(
        total_invested=total_invested_all,
        current_value=current_value_all,
        unrealized_gain=unrealized_all,
        unrealized_gain_pct=unrealized_pct_all,
        by_type=type_breakdown,
        by_platform=platform_breakdown,
        positions=positions,
    )

File: backend/app/services/reports/portfolio_service.rb
module Reports
  class PortfolioService
    # Quantities held in `decimal(12, 4)` — anything below this is rounding noise.
    QTY_EPSILON = 0.0001

    def initialize(user)
      @user = user
    end

    def call
      investments = @user.investments.includes(:platform_account, user_instrument: :instrument)
      return empty_result if investments.empty?

      positions = investments.group_by(&:user_instrument_id).map do |_ui_id, lots|
        instrument = lots.first.user_instrument&.instrument
        next nil unless instrument
        position_for(instrument, lots)
      end.compact.sort_by { |p| [ p[:is_closed] ? 1 : 0, -p[:current_value] ] }

      total_invested  = positions.sum { |p| p[:total_invested] }
      current_value   = positions.sum { |p| p[:current_value] }
      unrealized_gain = current_value - total_invested
      realized_gain   = positions.sum { |p| p[:realized_gain] }
      gain_pct        = total_invested > 0 ? (unrealized_gain / total_invested) * 100 : 0

      by_type = positions.group_by { |p| p[:type] }.map do |type, ps|
        {
          type:            type,
          investment_type: type,
          total_invested:  ps.sum { |p| p[:total_invested] },
          current_value:   ps.sum { |p| p[:current_value] },
          unrealized_gain: ps.sum { |p| p[:unrealized_gain] },
          realized_gain:   ps.sum { |p| p[:realized_gain] },
          count:           ps.size
        }
      end.sort_by { |h| -h[:current_value] }

      lots_by_platform = positions.flat_map { |p|
        p[:lots].map { |l| { platform: l[:platform_account_nickname] || "Unknown", position: p } }
      }
      by_platform = lots_by_platform.group_by { |x| x[:platform] }.map do |name, entries|
        positions_in_platform = entries.map { |e| e[:position] }.uniq
        invested = positions_in_platform.sum { |p| p[:total_invested] }
        current  = positions_in_platform.sum { |p| p[:current_value] }
        { platform_name: name, total_invested: invested, current_value: current }
      end.sort_by { |p| -p[:current_value] }

      {
        cost_basis_method:     "fifo",
        total_invested:        total_invested,
        current_value:         current_value,
        unrealized_gain:       unrealized_gain,
        unrealized_gain_pct:   gain_pct,
        realized_gain:         realized_gain,
        total_gain:            unrealized_gain + realized_gain,
        by_type:               by_type,
        by_platform:           by_platform,
        positions:             positions
      }
    end

    private

    def empty_result
      { cost_basis_method: "fifo", total_invested: 0, current_value: 0, unrealized_gain: 0, unrealized_gain_pct: 0,
        realized_gain: 0, total_gain: 0, by_type: [], by_platform: [], positions: [] }
    end

    def position_for(instrument, lots)
      investment_type             = lots.first.investment_type
      current_price, price_source = Holdings::PriceResolver.call(instrument, lots, investment_type)
      stats                       = Holdings::PositionCalculator.call(lots, current_price: current_price, investment_type: investment_type)

      cost_basis_held     = stats[:cost_basis_held]
      unrealized_gain     = stats[:unrealized_gain]
      unrealized_gain_pct = cost_basis_held > 0 ? (unrealized_gain / cost_basis_held) * 100 : 0.0

      platform_names = lots.map { |i| i.platform_account&.nickname }.compact.uniq

      {
        user_instrument_id:    lots.first.user_instrument_id,
        instrument_id:         instrument.id,
        instrument_name:       instrument.name,
        instrument_ticker:     instrument.ticker_symbol,
        instrument_exchange:   instrument.exchange,
        type:                  investment_type,
        platform_accounts:     platform_names,
        total_lots:            lots.count,
        buy_lots:              stats[:buy_lots],
        sell_lots:             stats[:sell_lots],
        is_closed:             stats[:is_closed],
        total_units:           stats[:total_units],
        long_term_units:       stats[:long_term_units],
        short_term_units:      stats[:short_term_units],
        total_invested:        cost_basis_held,
        net_cash_deployed:     stats[:net_cash_deployed],
        avg_buy_price:         stats[:avg_buy_price],
        current_price:         current_price.positive? ? current_price : nil,
        current_price_source:  price_source,
        current_price_at:      instrument.last_price_at,
        current_value:         stats[:current_value],
        unrealized_gain:       unrealized_gain,
        unrealized_gain_pct:   unrealized_gain_pct,
        realized_gain:         stats[:realized_gain],
        folio_number:          derive_folio_number(investment_type, lots),
        wavg:                  stats[:wavg],
        lots:                  lots.sort_by(&:purchase_date).map { |i| lot_json(i, stats[:lot_pnl]) }
      }
    end

    # Per-position folio number for mutual funds: latest buy lot's
    # folio_number, falling back to nil. Stocks don't have folios.
    def derive_folio_number(investment_type, lots)
      return nil unless investment_type == "mutual_fund"
      lots.select(&:buy?)
          .select { |l| l.folio_number.present? }
          .max_by(&:purchase_date)
          &.folio_number
    end

    def lot_json(inv, lot_pnl = {})
      {
        id:                          inv.id,
        trade_type:                  inv.trade_type,
        purchase_date:               inv.purchase_date,
        amount_invested:             inv.amount_invested.to_f,
        current_value:               inv.current_value&.to_f,
        quantity:                    inv.quantity&.to_f,
        units:                       inv.units&.to_f,
        price:                       inv.price&.to_f,
        folio_number:                inv.folio_number,
        platform_account_nickname:   inv.platform_account&.nickname,
        notes:                       inv.notes,
        pnl:                         lot_pnl[inv.id]
      }
    end
  end
end
